CP-27212: Update toolstack part when calling usb_reset.py

taoyongd opened the pull request:
According to deprivileged qemu design for usb passthrough part:
    1. before passing through USB device
    a) if it's the first USB device to pass through,
        1) bind mount /dev,/sys in chroot directory.
        2) create and join cgroup devices:/qemu-<domid>.
        3) blacklist all and add a minimal white list("c 1:3 r", allow read /dev/null) to cgroup allow list.
    b) change uid, gid of the device file.
    c) add device to cgroup allow list if no device of the same type is allowed.
    then xenopsd/device should
    d) call Device_add using QMP

    2. after stopping passing through USB device or the VM is shut down,
    the function vusb_unplug in xenopsd/device should
    a) call Device_del using QMP
    and then call a function vusb_cleanup in usb_reset.py to execute:
    b) restore uid, gid of the device file.
    c) remove device from cgroup allow list if no device of the same type is allowed.

    3. after QEMU dies (explicitly by xenopsd or implicitly by xenopsd detecting that the domain disappeared), the
    function QEMU.stop in xenopsd/device will call a function vm_cleanup in usb_reset.py to execute:
    a) remove the cgroup if one has been created.
    b) umount /dev, /sys from chroot directory ifthey are mounted.

This ticket is to refine toolstack part to call usb_reset.py correctly:
1. Refine the parameters passed to the caller in xenopsd.
2. QEMU will be deprivileged by default. But when pci-passthrough is enabled, then qemu will be privileged.
So vusb cleanup after stopping passing through USB device or the VM is shut down can be ignored if qemu is privileged.

Signed-off-by: Taoyong Ding <taoyong.ding@citrix.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/xapi-project/xenopsd/461)
<!-- Reviewable:end -->

lindig commented on xc/device.ml:
```diff
@@ -1409,29 +1409,42 @@ module Vkbd = struct
+      match action, priv with
+      | "attach", false ->
+        let stdout, stderr = Forkhelpers.execute_command_get_output Device_common.usb_reset_script [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid] in
+        debug "Got %s %s" stdout stderr
```
These debug lines make no sense when mixed in with other lines in the log because they don't provide enough context what this is about.

taoyongd replied:
updated

lindig commented on xc/device.ml:
```diff
@@ -1444,22 +1457,34 @@ module Vusb = struct
+        *)
+        begin match Qemu.pid ~xs domid with
+          | Some pid -> call_usb_reset hostbus hostport domid pid "attach" priv
+          | _ -> failwith "qemu pid does not exist"
```
Poor error message - provide more context - at least the domid.

taoyongd replied:
updated

lindig commented on xc/device.ml:
```diff
@@ -1444,22 +1457,34 @@ module Vusb = struct
+        (*Call vusb cleanup in usb_reset.py*)
+        match Qemu.pid ~xs domid with
+        | Some pid -> call_usb_reset hostbus hostport domid pid "detach" priv
+        | _ -> failwith "qemu pid does not exist"
```
The error message needs more context.

taoyongd replied:
updated

lindig commented on xc/device_common.ml:
```diff
@@ -334,6 +334,8 @@ let xenops_path_of_domain domid = sprintf "%s/%d" xenops_domain_path domid
 let xenops_vgpu_path domid devid =
   sprintf "%s/device/vgpu/%d" (xenops_path_of_domain domid) devid
 
+let usb_reset_script = "/opt/xensource/libexec/usb_reset.py"
```
This should be defined in `Xc_resources`

taoyongd replied:
updated

lindig commented on xc/device.ml:
```diff
@@ -1411,24 +1411,20 @@ end
       | "attach", false ->
-        let stdout, stderr = Forkhelpers.execute_command_get_output Device_common.usb_reset_script [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid] in
-        debug "Got %s %s" stdout stderr
+        ignore(Forkhelpers.execute_command_get_output !Xc_resources.usb_reset_script [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid])
```
The output would be actually useful for debugging. I suggest:

```
let cmd = [ "attach" ; ... ]  in (* please avoid long lines and break this up *)
let stdout, stderr = Forkhelpers.execute_command_get_output Device_common.usb_reset_script cmd in
debug "usb_reset_script '%s' returned stdout='%s' stderr='%s'" (String.concat " " cmd) stdout stderr
```
Since the script is called in several places, I would define a function and pass in just the arguments:

```
let exec_usb_reset_script argv =
  let stdout, stderr = Forkhelpers.execute_command_get_output Device_common.usb_reset_script argv in
  debug "usb_reset_script '%s' returned stdout='%s' stderr='%s'" (String.concat " " argv) stdout stderr
...
exec_usb_reset_script ["attach"; ..]
```

taoyongd replied:
Thanks. Updated. Have squashed the refine commit.

lindig commented on xc/device.ml:
```diff
@@ -1444,22 +1461,34 @@ module Vusb = struct
+        *)
+        begin match Qemu.pid ~xs domid with
+          | Some pid -> call_usb_reset hostbus hostport domid pid "attach" priv
+          | _ -> failwith (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid))
```
If this is raised, the subsequent `qmp` command `Device_add` will not be issued. Is this correct?

taoyongd replied:
If the pid does not exists, usb_reset.py shouldn't be called successfully. The pid is used when setting up cgroup, so I think the failure raise is correct and passthrough won't be success.

mseri replied:
Also, why we prefer `failwith` to something more tractable like `raise (Internal_error ...`

lindig commented on xc/xenops_server_xen.ml:
```diff
@@ -2035,20 +2035,33 @@ module VUSB = struct
        		   to release resources *)
     if not state.plugged then Some Needs_unplug else None
 
+  let is_privilege vm =
```
The name should be `is_privileged`

taoyongd commented on xc/device.ml:
```diff
@@ -1444,22 +1461,34 @@ module Vusb = struct
-    | _     -> ()
+    if is_running then
+      begin
+        qmp_send_cmd domid Qmp.(Device_del id) |> ignore;
```
Hi @lindig , if qmp_send_cmd is failed, seems the error will be raised and the vusb cleanup below won't be called. I have tested that when a vm is shut down with usb attached, I will catch an error "Caught (Unix.Unix_error "Connection refused" connect "")". IMO, when a vm is shut down, it calls vusb_unplug but the qmp socket is closed which raise a connection error. Am I right?

lindig replied:
If qmp_send fails, it will raise an error. If you need to call the cleanup action in any case, this should be wrapped in a `finally` call.

taoyongd replied:
If I unplug vusb firstly and then shut down vm, will not get the error.  What I mean here is that the uid/gid of the device file is not restored when catching the connection error, as the call_usb_reset is not called.

taoyongd replied:
OK. But we still can get the error in XenCenter. Is this ok?

lindig replied:
XenCenter receives its errors from xapi, which receives errors from xenopsd. If this is an unexpected error, this should be reported as `Internal_error` which will propagate to XenCenter. However, this should only be raised (or re-raised) after all cleanup actions have been performed.

mseri replied:
Also, if this error needs to have a proper error message in XenCenter, you need to define a new error in the idl, raise that and deal with it in xapi.

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
     with _ ->
-      error "Failed to reset USB device with hostbus %s and hostport %s." hostbus hostport;
+      error "Failed to %s USB device with hostbus %s and hostport %s." action hostbus hostport;
       failwith "Call to usb reset failed."
```
Why `failwith` and not something more tractable like `raise (Internal_error ...`?

mseri commented on xc/device.ml:
```diff
@@ -1444,22 +1461,34 @@ module Vusb = struct
+        (*Call vusb cleanup in usb_reset.py*)
+        match Qemu.pid ~xs domid with
+        | Some pid -> call_usb_reset hostbus hostport domid pid "detach" priv
+        | _ -> failwith (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid))
```
Here as well. I think we should use tractable exceptions in place of `failwith`s. It will be easier to deal with them at the xapi report level

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
+    try
+      let cmd = ["cleanup"; "-d"; string_of_int domid] in
+      exec_usb_reset_script cmd
+    with _ ->
```
Let's catch the error name and add it to the warning for now:
```ocaml
with err ->
   warn "Failed to clean up VM %s: %s" (string_of_int domid) (Printexc.to_string err)
```
If we see that it is useless for debugging purposes we can always change it later.

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
+        let cmd = [ "detach"; hostbus ^ "-" ^ hostport ; "-d"; string_of_int domid; "-i"; "0" ; "0"; ] in
+        exec_usb_reset_script cmd
+      | "detach", true -> () (*When qemu is privileged, detach will do nothing.*)
+      | _, _ -> () (*Invalid operation, will do nothing*)
```
Not even log that we sent an invalid operation?

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
+        exec_usb_reset_script cmd
+      | "detach", true -> () (*When qemu is privileged, detach will do nothing.*)
+      | _, _ -> () (*Invalid operation, will do nothing*)
     with _ ->
```
Should we log here what kind of error? Hiding the errors is often making debugging miserable

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
-    | _ -> ()
 
-  let vusb_plug ~xs ~domid ~id ~hostbus ~hostport ~version =
+  let vusb_plug ~xs ~priv ~domid ~id ~hostbus ~hostport ~version =
```
let's call it `privileged` and make the code more readable

mseri commented on xc/device.ml:
```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
+    with _ ->
+      warn "Failed to clean up VM %s" (string_of_int domid)
+
+  let call_usb_reset hostbus hostport domid pid action priv =
```
Make the parameter into named parameter, these are all strings at least we can use the names to prevent ordering errors

mseri commented on xc/device.ml:
```diff
@@ -2227,7 +2256,9 @@ module Backend = struct
           (qmp_event_path domid);
           (qmp_libxl_path domid);
-        ] |> List.iter Socket.Unix.rm
+        ] |> List.iter Socket.Unix.rm;
```
Is this safe or could throw an exception and prevent the cleanup below?

lindig replied:
These are simple functions that just return the path - I think it is safe

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[xc/device.ml, line 1423 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1423:-L6VCqyin_RRE_YQoDfV:bmesf68) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1423)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Make the parameter into named parameter, these are all strings, so at least we can use the names to prevent ordering errors
</blockquote></details>

Ok, Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[xc/device.ml, line 1420 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1420:-L6VEayKJRMvLeIT0Tu0:bjhrbs9) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1420)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Let's catch the error name and add it to the warning for now:
```ocaml
with err ->
   warn "Failed to clean up VM %s: %s" (string_of_int domid) (Printexc.to_string err)
```
If we see that it is useless for debugging purposes we can always change it later.
</blockquote></details>

Yes, updated.

---

*[xc/device.ml, line 1437 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1437:-L6VEnMGeFIa9mSuxmNy:b-6svz2q) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1437)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Not even log that we sent an invalid operation?
</blockquote></details>

Have refined the usb_reset call function which will be much easier than this.

---

*[xc/device.ml, line 1438 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1438:-L6VF0VOar21fWBlablC:b-frmxgy) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1438)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Should we log here what kind of error? Hiding the errors is often making debugging miserable
</blockquote></details>


Have refined the usb_reset call function which will be much easier than this.

---

*[xc/device.ml, line 1440 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1440:-L6VF5TBbDd3fMAHbsl-:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1440)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Why `failwith` and not something more tractable like `raise (Internal_error ...`?
</blockquote></details>

Updated.

---

*[xc/device.ml, line 1447 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1447:-L6VE8UN3clUOEyX9W6W:b1wjbw5) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1447)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

let's call it `privileged` and make the code more readable
</blockquote></details>

Thanks, updated.

---

*[xc/device.ml, line 1471 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1471:-L6VF8KNz3f2sV_mD_3j:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1471)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Also, why we prefer `failwith` to something more tractable like `raise (Internal_error ...`
</blockquote></details>

Updated.

---

*[xc/device.ml, line 1486 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1486:-L6VEG-ZAJv_rr3JOIYp:b1x85p4) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1486)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Also, if this error needs to have a proper error message in XenCenter, you need to define a new error in the idl, raise that and deal with it in xapi (in addition to update xencenter to know the error).
</blockquote></details>

Yes. For this case I described, I will raise another PR the to discuss the error, so will not continue on this PR. Thanks.

---

*[xc/xenops_server_xen.ml, line 2038 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWL-r2-2038:-L6VDiPoQ_XAIFHRnurm:b1wjbw5) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/xenops_server_xen.ml#L2038)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

The name should be `is_privileged`
</blockquote></details>

Thanks, updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[xc/device.ml, line 1490 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-1490:-L6VFJhCq8vD3WPGH1m1:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L1490)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Here as well. I think we should use tractable exceptions in place of `failwith`s. It will be easier to deal with them at the xapi report level
</blockquote></details>

Updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

mseri commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[xc/device.ml, line 1450 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VtzvtCGyWNOqNs6-y:-L6VtzvuHcqXyY-pL5V6:bgalb9i) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1450)):*
> ```OCaml
>     if device_model = Profile.Qemu_trad then
>       raise (Internal_error (Printf.sprintf "Failed to plug VUSB %s because domain %d uses device-model profile %s." id domid (Profile.string_of device_model)));
>     debug "Vusb plugg: plug vusb device %s" id;
> ```

Let's use the function name `"vusb_plug: plug VUSB device %s" id;`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

mseri commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[xc/device.ml, line 1433 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VvBrAIjh_kUu2JHxg:-L6VvBrAIjh_kUu2JHxh:b-p62hp0) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1433)):*
> ```OCaml
>       | true ->
>         [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid; "-r" ]
>     in
> ```

Given that most of the args are in common, maybe it's better to de-dupe them. Something on the line of

```ocaml
let argv =  List.concat [
      [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid ]
    , if privileged then ["-r"] else []
    ]
in
```
but, to be honest, I don't have a strong opinion on either approach in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[xc/device.ml, line 1433 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VvBrAIjh_kUu2JHxg:-L6Vx8aCbibrD_gnP3fC:bsjr2dv) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1433)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Given that most of the args are in common, maybe it's better to de-dupe them. Something on the line of

```ocaml
let argv =  List.concat [
      [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid ]
    , if privileged then ["-r"] else []
    ]
in
```
but, to be honest, I don't have a strong opinion on either approach in this case.
</blockquote></details>

OK. I think de-dupe will be better. 

---

*[xc/device.ml, line 1450 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VtzvtCGyWNOqNs6-y:-L6VxSHI5LzdbviQp-wJ:b2nhll9) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1450)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

Let's use the function name `"vusb_plug: plug VUSB device %s" id;`
</blockquote></details>

Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[xc/device.ml, line 1433 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VvBrAIjh_kUu2JHxg:-L6W2408u9Hd8ctC6dGP:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1433)):*
<details><summary><i>Previously, taoyongd (taoyongd) wrote…</i></summary><blockquote>

OK. I think de-dupe will be better. 
</blockquote></details>

Updated.

---

*[xc/device.ml, line 1450 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VtzvtCGyWNOqNs6-y:-L6W25b9qoURtU8Apc5C:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1450)):*
<details><summary><i>Previously, taoyongd (taoyongd) wrote…</i></summary><blockquote>

Ok.
</blockquote></details>

Updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
@lindig  @mseri, any comments  about this PR?

lindig commented:
*[xc/device.ml, line 1450 at r3](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VtzvtCGyWNOqNs6-y:-L6W6PCOAndThYR4wCDX:b76xlpf) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1450)):*
<details><summary><i>Previously, taoyongd (taoyongd) wrote…</i></summary><blockquote>

Updated.
</blockquote></details>

Please avoid overlong lines in new code. You can see here that this makes them difficult  to read. If you need space for long argument lists, do:
```
[ ...
; ...
; sprintf "%s-%s" hostbus hostport
]
```

I also think that using sprintf is better than "^" except in the simplest cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

lindig commented:
*[xc/device.ml, line 2259 at r2](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-2259:-L6W7DYHBJssQncNXolV:b-5yi5mi) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L2259)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

These are simple functions that just return the path - I think it is safe
</blockquote></details>

I have to correct myself. This needs to be protected against failure:

```
let rm path =
  try Socket.Unix.rm path
  with e -> error "rm %s failed: %s (%s)" path (Printexec.to_string e) __LOC__ 
in
...
...List.iter rm
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

lindig commented:
*[xc/device_common.ml, line 337 at r1](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6W81b9bfB6cO51H_V7:-L6W81b9bfB6cO51H_V8:b-oaqs1k) ([raw file](https://github.com/xapi-project/xenopsd/blob/fd23c1018b3fe9a734ea4e5210ff072684043746/xc/device_common.ml#L337)):*
> ```OCaml
>   sprintf "%s/device/vgpu/%d" (xenops_path_of_domain domid) devid
> 
> let usb_reset_script = "/opt/xensource/libexec/usb_reset.py"
> ```

I think this should be defined in `xc_resources.ml`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[xc/device_common.ml, line 337 at r1](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-L6W81b9bfB6cO51H_V7:-L6WAj-3NYe5MVOoN--N:benr346) ([raw file](https://github.com/xapi-project/xenopsd/blob/fd23c1018b3fe9a734ea4e5210ff072684043746/xc/device_common.ml#L337)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

I think this should be defined in `xc_resources.ml`
</blockquote></details>

Hi @lindig, this is already in xc_resources.ml.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[xc/device.ml, line 1450 at r3](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6VtzvtCGyWNOqNs6-y:-L6WDva49yewQ2CXykO9:b-tqdpx5) ([raw file](https://github.com/xapi-project/xenopsd/blob/7cd40abecadd2846006c7ae8aab212eae3f7a89e/xc/device.ml#L1450)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

Please avoid overlong lines in new code. You can see here that this makes them difficult  to read. If you need space for long argument lists, do:
```
[ ...
; ...
; sprintf "%s-%s" hostbus hostport
]
```

I also think that using sprintf is better than "^" except in the simplest cases.
</blockquote></details>

OK. Updated with such case.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

lindig commented:
Getting there. Some minor comments.

---







*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-:-L6XbKjV8cdLU2tUgF5L:b-8x8aum)*
<!-- Sent from Reviewable.io -->

lindig commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[xc/device.ml, line 1483 at r5](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6XaWLR1bEJ5Pjv41sv:-L6XaWLR1bEJ5Pjv41sw:b1ew3ur) ([raw file](https://github.com/xapi-project/xenopsd/blob/306a27c025debad6dbec33d5eff831433d20d907/xc/device.ml#L1483)):*
> ```OCaml
>         begin match Qemu.pid ~xs domid with
>           | Some pid -> usb_reset_attach ~hostbus ~hostport ~domid ~pid ~privileged
>           | _ -> raise (Internal_error (Printf.sprintf "qemu pid does not exist for vm %s" (string_of_int domid)))
> ```

Use `sprintf ".... domid %d" domoid` and don't use `string_of_int` 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[xc/device.ml, line 1483 at r5](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6XaWLR1bEJ5Pjv41sv:-L6ZXEpR7fO7xeDeiZrw:b1wjbw5) ([raw file](https://github.com/xapi-project/xenopsd/blob/306a27c025debad6dbec33d5eff831433d20d907/xc/device.ml#L1483)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

Use `sprintf ".... domid %d" domoid` and don't use `string_of_int` 
</blockquote></details>

Thanks, updated.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Thanks, updated.

---

Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-:-L6ZXQ0R-F1xqmffHWuY:b1wjbw5)*
<!-- Sent from Reviewable.io -->

mseri commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[xc/device.ml, line 2259 at r2](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-2259:-L6a131H4m69sm0jzhAn:b-x1y8y6) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L2259)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

I have to correct myself. This needs to be protected against failure:

```
let rm path =
  try Socket.Unix.rm path
  with e -> error "rm %s failed: %s (%s)" path (Printexec.to_string e) __LOC__ 
in
...
...List.iter rm
```
</blockquote></details>

This needs to be updated as well, I don't seem to see the code changes in the last commit

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

lindig commented:
*[xc/device.ml, line 2259 at r2](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-2259:-L6a1iqJ8332mKVgLY3q:bj5ampe) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L2259)):*
<details><summary><i>Previously, mseri (Marcello Seri) wrote…</i></summary><blockquote>

This needs to be updated as well, I don't seem to see the code changes in the last commit
</blockquote></details>

To be fair, this code existed before but please fix it.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[xc/device.ml, line 2259 at r2](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-2259:-L6a8QGg1IYJhFBf0RM1:b99o6l6) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L2259)):*
<details><summary><i>Previously, lindig (Christian Lindig) wrote…</i></summary><blockquote>

To be fair, this code existed before but please fix it.
</blockquote></details>

Ok, will update it also.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[xc/device.ml, line 2259 at r2](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-L6VAcbGCnE5NQq1rSWH-r2-2259:-L6aBb0V68hfhoulSY0w:b-bw8e14) ([raw file](https://github.com/xapi-project/xenopsd/blob/76addafec985c514e76ec63444b6dc4e2b5b5ab5/xc/device.ml#L2259)):*
<details><summary><i>Previously, taoyongd (taoyongd) wrote…</i></summary><blockquote>

Ok, will update it also.
</blockquote></details>

Updated.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Any comments on this? @lindig @mersi

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-:-L6aE62FCX941NBcK2gt:bu9laa4)*
<!-- Sent from Reviewable.io -->

taoyongd commented:
Sorry. mseri not mersi.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xenopsd/461#-:-L6aGTIy2qlNn0-fjUq2:b-mdov5)*
<!-- Sent from Reviewable.io -->

mseri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 5 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-:-L6aJKQ72NkS6418LJrT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mseri commented:
@taoyongd I think by rebasing you got lots of changes from other PRs or branches. Is that expected?

taoyongd commented:
I rebase the qemu-upstream branch which is what i need to when it is merged.  Isn't this correct?

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xenopsd/461#-:-L6aUTfUwfYYwn286O71:b-v34vgp)*
<!-- Sent from Reviewable.io -->

mseri commented:
I just wanted to duble check, thanks. If @lindig is happy as well, we can merge

taoyongd commented:
Ok. Got it. Thanks and sorry for the rebase ahead. :)

lindig commented:
We are in the process of rebasing the qemu branch. Please wait until we completed that and rebase on top of that.

lindig commented:
We have updated the xenopsd branch. Please rebase your changes.

taoyongd commented:
Thanks @lindig . I have rebased the branch.

taoyongd commented:
Hi @lindig , can this PR now avaliable to be merged? Thanks

taoyongd commented:
Hi @lindig, I have checked the indentation with ocp-indent which looks ok. The long line logs are short now. About the comment in vusb_unplug, it seems no need anymore as more info given in `exec_usb_reset_script`, so removed.
